### DISCUSSION TITLE
compiling in Cygwin?

### DISCUSSION THREAD

**Discussion by luppahannu** [stefanhaustein/TerminalImageViewer#125] [Q&A] [answered]
Hey!
I have used tiv under Linux, and now tried compiling under Cygwin, but no luck. Make spews out lines and lines of warnings and errors, and eventually exits with an error.

Is this doable? Hard to tell if there is some package that I am missing. ImageMagick is installed.

**Comment by aaronliu0130** [comment, marked as answer]
Hey! This is the first actual discussion in this repo! 🎉
std::filesystem requires c++17+, have you passed the necessary arguments? If I remember right cygwin uses a version of g++ compiled for windows, which should use gnu++14 by default.
The only dependency is imagemagick. If it spits out lines like `\033[1;31m` then it means that your terminal doesn't support linux color codes. 
If this still doesn't solve your question, could I see a screenshot?

EDIT: answer is below from op

**Reply by luppahannu** [reply]
Well... I just run make, and it seems to pass the correct -std=c++17 argument. The whole line is g++ -O2 -fpermissive -std=c++17 -Wall -fexceptions   -c -o tiv.o tiv.cpp

I directed the output of the compiling to a text file and it was over 200k, over 3000 lines. I don't want to spam by attaching the whole thing, I'll put a few lines from the beginning and the end

<details><summary>Beginning</summary>
<pre>
g++ -O2 -fpermissive -std=c++17 -Wall -fexceptions   -c -o tiv.o tiv.cpp
In file included from tiv.cpp:13:
CImg.h: In member function ‘cimg_library::CImg<T>& cimg_library::CImg<T>::load_graphicsmagick_external(const char*)’:
CImg.h:50853:13: warning: there are no arguments to ‘pclose’ that depend on a template parameter, so a declaration of ‘pclose’ must be available [-fpermissive]
50853 |             pclose(file);
      |             ^~~~~~
CImg.h:50861:11: warning: there are no arguments to ‘pclose’ that depend on a template parameter, so a declaration of ‘pclose’ must be available [-fpermissive]
50861 |           pclose(file);
      |           ^~~~~~
CImg.h: In member function ‘cimg_library::CImg<T>& cimg_library::CImg<T>::load_imagemagick_external(const char*)’:
CImg.h:50970:13: warning: there are no arguments to ‘pclose’ that depend on a template parameter, so a declaration of ‘pclose’ must be available [-fpermissive]
50970 |             pclose(file);
      |             ^~~~~~
CImg.h:50978:11: warning: there are no arguments to ‘pclose’ that depend on a template parameter, so a declaration of ‘pclose’ must be available [-fpermissive]
50978 |           pclose(file);
      |           ^~~~~~
CImg.h: In member function ‘cimg_library::CImg<T>& cimg_library::CImg<T>::load_dcraw_external(const char*)’:
CImg.h:51083:11: warning: there are no arguments to ‘pclose’ that depend on a template parameter, so a declaration of ‘pclose’ must be available [-fpermissive]
51083 |           pclose(file);
      |           ^~~~~~
CImg.h:51090:9: warning: there are no arguments to ‘pclose’ that depend on a template parameter, so a declaration of ‘pclose’ must be available [-fpermissive]
51090 |         pclose(file);
      |         ^~~~~~
CImg.h: In member function ‘const cimg_library::CImg<T>& cimg_library::CImg<T>::_save_pandore(FILE*, const char*, unsigned int) const’:
CImg.h:54034:11: warning: this ‘for’ clause does not guard... [-Wmisleading-indentation]
54034 |           for (int d = 0; d<5; ++d) ndims[d] = (unsigned long)dims[d]; cimg::fwrite(ndims._data,nbdims,nfile); } \
      |           ^~~
</pre>
</details>

<details><summary>Ending</summary>
<pre>
CImg.h: In instantiation of ‘cimg_library::CImg<T>& cimg_library::CImg<T>::load_graphicsmagick_external(const char*) [with T = unsigned char]’:
CImg.h:51219:17:   required from ‘cimg_library::CImg<T>& cimg_library::CImg<T>::load_other(const char*) [with T = unsigned char]’
CImg.h:48195:11:   required from ‘cimg_library::CImg<T>& cimg_library::CImg<T>::load(const char*) [with T = unsigned char]’
CImg.h:11490:14:   required from ‘cimg_library::CImg<T>& cimg_library::CImg<T>::assign(const char*) [with T = unsigned char]’
CImg.h:11137:13:   required from ‘cimg_library::CImg<T>::CImg(const char*) [with T = unsigned char]’
tiv.cpp:510:51:   required from here
CImg.h:50848:21: error: ‘popen’ was not declared in this scope; did you mean ‘fopen’?
50848 |         file = popen(command,"r");
      |                ~~~~~^~~~~~~~~~~~~
      |                fopen
CImg.h:50853:19: error: ‘pclose’ was not declared in this scope; did you mean ‘fclose’?
50853 |             pclose(file);
      |             ~~~~~~^~~~~~
      |             fclose
CImg.h:50861:17: error: ‘pclose’ was not declared in this scope; did you mean ‘fclose’?
50861 |           pclose(file);
      |           ~~~~~~^~~~~~
      |           fclose
CImg.h: In instantiation of ‘static const cimg_library::CImgList<typename cimg_library::cimg::last<T, unsigned char>::type>& cimg_library::CImgList<T>::font(unsigned int, bool) [with T = unsigned char; typename cimg_library::cimg::last<T, unsigned char>::type = unsigned char]’:
CImg.h:44627:60:   required from ‘cimg_library::CImg<T>& cimg_library::CImg<T>::draw_text(int, int, const char*, const tc1*, const tc2*, float, unsigned int, ...) [with tc1 = unsigned char; tc2 = unsigned char; T = unsigned char]’
CImg.h:47224:33:   required from ‘cimg_library::CImg<typename cimg_library::cimg::last<T, int>::type> cimg_library::CImg<T>::_select(cimg_library::CImgDisplay&, const char*, unsigned int, unsigned int*, int, int, int, bool, bool, bool, bool) const [with T = double; typename cimg_library::cimg::last<T, int>::type = int]’
CImg.h:51402:37:   required from ‘const cimg_library::CImg<T>& cimg_library::CImg<T>::_display(cimg_library::CImgDisplay&, const char*, bool, unsigned int*, bool, bool) const [with T = double]’
CImg.h:51340:14:   required from ‘const cimg_library::CImg<T>& cimg_library::CImg<T>::display(const char*, bool, unsigned int*, bool) const [with T = double]’
CImg.h:20675:23:   required from ‘static double cimg_library::CImg<T>::_cimg_math_parser::mp_display_memory(cimg_library::CImg<T>::_cimg_math_parser&) [with T = float]’
CImg.h:17853:38:   required from ‘unsigned int cimg_library::CImg<T>::_cimg_math_parser::compile(char*, char*, unsigned int, unsigned int*, bool) [with T = float]’
CImg.h:15281:41:   required from ‘cimg_library::CImg<T>::_cimg_math_parser::_cimg_math_parser(const char*, const char*, const cimg_library::CImg<T>&, cimg_library::CImg<T>*, const cimg_library::CImgList<T>*, cimg_library::CImgList<T>*, bool) [with T = float]’
CImg.h:24629:25:   required from ‘double cimg_library::CImg<T>::_eval(cimg_library::CImg<T>*, const char*, double, double, double, double, const cimg_library::CImgList<T>*, cimg_library::CImgList<T>*) const [with T = float]’
CImg.h:24615:19:   required from ‘double cimg_library::CImg<T>::eval(const char*, double, double, double, double, const cimg_library::CImgList<T>*, cimg_library::CImgList<T>*) const [with T = float]’
CImg.h:60883:22:   required from here
CImg.h:59570:21: warning: ‘void* memmove(void*, const void*, size_t)’ writing to an object of type ‘struct cimg_library::CImgList<unsigned char>’ with no trivial copy-assignment; use copy-assignment or copy-initialization instead [-Wclass-memaccess]
59570 |         std::memmove(fonts,fonts + 1,15*sizeof(CImgList<ucharT>));
      |         ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
CImg.h:54682:10: note: ‘struct cimg_library::CImgList<unsigned char>’ declared here
54682 |   struct CImgList {
      |          ^~~~~~~~
make: *** [<builtin>: tiv.o] Error 1
</pre>
</details>

Oh boy github does not wan't to do that as easy as other forum software :smile:

**Reply by aaronliu0130** [reply]
Popen and pclose are Linux/Unix only commands, do they work under Cygwin?

I'll also take a look to see if the newer version of cimg still has this issue.

**Reply by luppahannu** [reply]
I copied a small popen-test code from somewhere, and tried compiling it. 

Without any extra parameters, under Cygwin g++ compiles it without any errors and the executable runs correctly. But when I put -std=c++17 (or seems like any other c++ version)  the errors start coming out, it fails to see what popen is. 
I tried the same stuff on my linux machine and everything compiles just fine.

After that I tried -std=gnu++17 under Cygwin, and that compiles and the executable works. Then I modified the tiv Makefile and replaced 
`override CXXFLAGS += -std=c++17 -Wall -fexceptions`
with
`override CXXFLAGS += -std=gnu++17 -Wall -fexceptions`

and now running make works! The program compiles and it runs, and I can see the pictures.

**Reply by aaronliu0130** [reply]
Glad you managed to solve it!
